src/sqlfluff/core/config.py

```diff
@@ -144,6 +146,31 @@ def get_global(cls) -> "ConfigLoader":
             global_loader = cls()
         return global_loader
 
+    @classmethod
+    def _walk_toml(cls, config: Dict[str, Any], base_key=()):
+        """Recursively walk the nested config inside a TOML file."""
+        buff: List[tuple] = []
+        for k, v in config.items():
+            key = base_key + (k,)
+            if isinstance(v, dict):
+                buff.extend(cls._walk_toml(v, key))
+            else:
+                buff.append((key, v))
+
+        return buff
+
+    @classmethod
+    def _get_config_elems_from_toml(cls, fpath: str) -> List[Tuple[tuple, Any]]:
+        """Load a config from a TOML file and return a list of tuples.
+
+        The return value is a list of tuples, were each tuple has two elements,
+        the first is a tuple of paths, the second is the value at that path.
+        """
+        config = toml.load(fpath)
+        tool = config.get("tool", {}).get("sqlfluff", {})
```

Comment:
Now that you replaced `__options` with the `tool.sqlfluff.core` technique, do you need to move or place the `core` values to match the structure of the other config file types?

In other words, have you verified that SQLFluff correctly uses the values from the `core` section?
```
[tool.sqlfluff.core]
    templater = "jinja"
    sql_file_exts = [
        ".sql",
        ".sql.j2",
        ".dml",
        ".ddl",
    ]
```

Reply:
@barrywhart I added a test case with

```toml
[tool.sqlfluff.core]
testing_int = 5
testing_bar = 7.698
testing_bool = false
testing_arr = [ "a", "b", "c" ]
testing_inline_table = { x = 1 }
 
[tool.sqlfluff.bar]
foo = "foobar"

[tool.sqlfluff.fnarr.fnarr]
foo = "foobar"
```

which gets collected into the following dict:

```python
{
    "core": {
        "testing_int": 5,
        "testing_bar": 7.698,
        "testing_bool": False,
        "testing_arr": ["a", "b", "c"],
        "testing_inline_table": {"x": 1},
    },
    "bar": {"foo": "foobar"},
    "fnarr": {"fnarr": {"foo": "foobar"}},
}
```

does it look right?``

Reply:
TBH, I'd feel much better if you could do some manual tests to confirm the settings in the TOML file are being read correctly and actually affecting the runtime behavior of SQLFluff. There are sometimes limits to the effectiveness of automated tests and manual review, and I think this is one of those times.

Reply:
@barrywhart  That makes a lot of sense. I created [this repo](https://github.com/edgarrmondragon/sqlfluff-pyproject) to test some configurations. Let me know if you think there are edge cases I could add.

Reply:
I tested with your example repo. Works great! I also added a simple test with templating. This is based on the Variable Templating [example](https://docs.sqlfluff.com/en/stable/configuration.html#variable-templating) from the SQLFluff docs.

`test.sql`:
```
SELECT {{ num_things }} FROM {{ tbl_name }} WHERE id > 10 LIMIT 5
```

Additions to `pyproject.toml`:
```
[tool.sqlfluff.core]
templater = "jinja"

[tool.sqlfluff.templater.jinja.context]
num_things=456
tbl_name="my_table"
```

Reply:
I added that templated example to the example repo :smile:.